File: apps/androidtv/player/engine-mpv/src/main/kotlin/dev/spidola/tv/player/enginempv/MpvEngine.kt
// SPDX-FileCopyrightText: 2026 Spidola contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

package dev.spidola.tv.player.enginempv

import android.view.SurfaceHolder
import android.view.SurfaceView
import androidx.compose.foundation.layout.Box
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.viewinterop.AndroidView
import dev.spidola.tv.core.playercontract.AspectMode
import dev.spidola.tv.core.playercontract.BufferingProfile
import dev.spidola.tv.core.playercontract.EngineError
import dev.spidola.tv.core.playercontract.EngineId
import dev.spidola.tv.core.playercontract.PlaybackEngine
import dev.spidola.tv.core.playercontract.PlaybackState
import dev.spidola.tv.core.playercontract.StreamRequest
import dev.spidola.tv.core.playercontract.TrackId
import dev.spidola.tv.core.playercontract.TrackSelection
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.concurrent.atomic.AtomicBoolean

/**
 * The libmpv engine — mpv-class codec breadth, the Android fallback (TECH_SPEC §8).
 *
 * Construction is free of I/O and touches no mpv: the zap path destroys and rebuilds an
 * engine per channel flip, so everything expensive waits for [load].
 *
 * Threading: the contract makes engines main-thread-affine, and mpv's event pump is not.
 * Events arrive on [MpvClient]'s pump thread and are published straight into
 * [MutableStateFlow]s, which are safe to write from any thread. That is deliberately all the
 * concurrency machinery here — no dispatcher, no scope this engine would then have to own
 * and cancel, and no `runBlocking`.
 */
class MpvEngine : PlaybackEngine {
    override val id: EngineId = EngineId.MPV

    private val _state = MutableStateFlow<PlaybackState>(PlaybackState.Idle)
    override val state: StateFlow<PlaybackState> = _state.asStateFlow()

    private val _tracks = MutableStateFlow(TrackSelection())
    override val tracks: StateFlow<TrackSelection> = _tracks.asStateFlow()

    private val _isSeekable = MutableStateFlow(false)
    override val isSeekable: StateFlow<Boolean> = _isSeekable.asStateFlow()

    private val released = AtomicBoolean(false)

    @Volatile
    private var client: MpvClient? = null

    /**
     * The last few lines mpv logged, kept only long enough to classify a failure.
     * [MpvErrorMapping] needs them because mpv's error codes cannot tell 401 from a DNS
     * failure. Bounded, and redacted on the way in.
     */
    private val recentDiagnostics = MpvDiagnosticBuffer()

    @Volatile
    private var flags = MpvStateDerivation.Flags()

    /**
     * Serialises surface attach/detach against each other and against teardown. The
     * composable disposes on the main thread while the pump runs elsewhere, and mpv's VO
     * teardown must not interleave with a new attach.
     */
    private val surfaceLock = Any()
    private var attachedSurface: android.view.Surface? = null

    @Composable
    override fun Surface(modifier: Modifier) {
        // SurfaceView, not TextureView. On TV this is not a close call: a SurfaceView gets a
        // dedicated hardware overlay plane, so video bypasses view compositing entirely.
        // TextureView routes every frame through GPU composition, which costs power and
        // rules out HDR passthrough — on exactly the low-end Chromecast-class device PRD §9
        // sets as the baseline.
        Box(modifier) {
            AndroidView(
                factory = { context ->
                    SurfaceView(context).apply {
                        holder.addCallback(
                            object : SurfaceHolder.Callback {
                                override fun surfaceCreated(holder: SurfaceHolder) {
                                    onSurfaceAvailable(holder.surface)
                                }

                                override fun surfaceChanged(
                                    holder: SurfaceHolder,
                                    format: Int,
                                    width: Int,
                                    height: Int,
                                ) {
                                    // mpv reads the window size from the ANativeWindow itself
                                    // (android_common.c:84-96). Telling it again would only
                                    // race that read.
                                }

                                override fun surfaceDestroyed(holder: SurfaceHolder) {
                                    onSurfaceLost()
                                }
                            },
                        )
                    }
                },
            )
        }
    }

    override fun load(request: StreamRequest) {
        if (released.get()) return
        check(client == null) { "load() called twice; engines are single-use — rebuild instead" }

        MpvLog.loading(request)
        publish(PlaybackState.Loading)

        val mpv = MpvClient.create()
        if (mpv == null) {
            // No libmpv in this build. An honest terminal failure beats an
            // UnsatisfiedLinkError escaping into a composable.
            publish(PlaybackState.Failed(EngineError.Unknown("libmpv is not available in this build")))
            return
        }
        client = mpv

        applyPreInitOptions(mpv, request)

        val initRc = mpv.initialize()
        if (initRc < 0) {
            publish(PlaybackState.Failed(MpvErrorMapping.engineErrorFor(initRc, "mpv_initialize failed")))
            return
        }

        // Warnings and worse only. Not merely to keep logcat quiet: mpv's info level echoes
        // the options it resolved, which is where a token-bearing header would be printed.
        // MpvLogRedaction is the backstop, not the first line of defence.
        mpv.requestLogMessages("warn")
        observeDefaults(mpv)

        // The composable may have produced a surface before load() ran — there is no ordering
        // guarantee between them, so whichever arrives second does the attach.
        synchronized(surfaceLock) {
            attachedSurface?.let { attach(mpv, it) }
        }

        mpv.startEventPump(::onEvent)

        // Everything above is configuration. This is the line that starts work.
        val loadRc = mpv.command("loadfile", request.locator)
        if (loadRc < 0) {
            publish(PlaybackState.Failed(MpvErrorMapping.engineErrorFor(loadRc, recentDiagnostics.snapshot())))
        }
    }

    override fun play() {
        client?.setProperty("pause", "no")
    }

    override fun pause() {
        client?.setProperty("pause", "yes")
    }

    override fun seekTo(seconds: Double) {
        // A no-op when not seekable, per the contract, so the caller need not guard.
        if (!_isSeekable.value) return
        client?.command("seek", seconds.toString(), "absolute")
    }

    override fun select(track: TrackId) {
        val selection = MpvTrackMapping.selectionFor(track) ?: return
        client?.setProperty(selection.first, selection.second)
    }

    override fun clearSubtitle() {
        // "no" is mpv's off value for a track property. Distinct from selecting a track,
        // because "no subtitle" is not a track — which is why the contract has this method.
        client?.setProperty("sid", "no")
    }

    override fun setAspect(mode: AspectMode) {
        val mpv = client ?: return
        when (mode) {
            // keepaspect + panscan=0: letterbox, whole frame visible.
            AspectMode.FIT -> {
                mpv.setProperty("keepaspect", "yes")
                mpv.setProperty("panscan", "0")
            }
            // panscan=1 zooms until the frame fills the screen, cropping the overflow.
            AspectMode.FILL -> {
                mpv.setProperty("keepaspect", "yes")
                mpv.setProperty("panscan", "1")
            }
            // keepaspect=no stretches to the window, ignoring the source ratio.
            AspectMode.STRETCH -> {
                mpv.setProperty("keepaspect", "no")
                mpv.setProperty("panscan", "0")
            }
        }
    }

    override fun release() {
        if (!released.compareAndSet(false, true)) return
        val mpv = client ?: return

        // Surface first, then mpv: detachSurface's contract is that the VO is already down.
        synchronized(surfaceLock) {
            if (attachedSurface != null) {
                mpv.setProperty("vo", "null")
                mpv.detachSurface()
                attachedSurface = null
            }
        }

        // Joins the pump, then destroys the handle.
        mpv.release()
        client = null
    }

    // region Surface ownership

    private fun onSurfaceAvailable(surface: android.view.Surface) {
        synchronized(surfaceLock) {
            attachedSurface = surface
            // load() may not have run yet; it re-attaches from attachedSurface if so.
            client?.let { attach(it, surface) }
        }
    }

    private fun onSurfaceLost() {
        synchronized(surfaceLock) {
            val mpv = client
            if (mpv != null && attachedSurface != null) {
                // The ordering that prevents the use-after-free. mpv's render thread holds an
                // ANativeWindow derived from this Surface; `vo` is a property, so setting it
                // is synchronous, and when it returns the VO is down and the window released.
                // Only then may the global ref go.
                //
                // surfaceDestroyed must not return until nothing will touch the surface
                // again — the system frees it immediately after. So blocking here is the
                // contract, not an oversight.
                mpv.setProperty("vo", "null")
                mpv.detachSurface()
            }
            attachedSurface = null
        }
    }

    private fun attach(
        mpv: MpvClient,
        surface: android.view.Surface,
    ) {
        val rc = mpv.attachSurface(surface)
        if (rc < 0) {
            MpvLog.nativeCallFailed("attachSurface", rc)
            return
        }
        // vo comes up only after wid is set: mpv reads wid when the VO initialises, so the
        // reverse order hands it a null window.
        mpv.setProperty("vo", "gpu")
    }

    // endregion

    private fun applyPreInitOptions(
        mpv: MpvClient,
        request: StreamRequest,
    ) {
        // vo starts null and is switched on when a surface arrives; otherwise mpv initialises
        // video output against a window that does not exist yet.
        mpv.setOption("vo", "null")
        mpv.setOption("gpu-context", "android")
        mpv.setOption("ao", "audiotrack,opensles")
        // This is an embedded library: mpv's own config and scripts must never load, or a
        // stray config on a rooted device would change playback unreproducibly.
        mpv.setOption("config", "no")
        mpv.setOption("osc", "no")
        mpv.setOption("input-default-bindings", "no")
        // MediaCodec hardware decode, falling back to software. `-copy` reads frames back for
        // the GL renderer: it costs memory bandwidth, but direct rendering needs a surface
        // mpv owns outright, which is incompatible with handing it ours.
        mpv.setOption("hwdec", "mediacodec-copy,no")

        applyBuffering(mpv, request.buffering)
        applyHeaders(mpv, request)
    }

    private fun applyBuffering(
        mpv: MpvClient,
        profile: BufferingProfile,
    ) {
        // Honest accounting: these are starting points, not measured optima. The trade is
        // real — cache-secs is roughly how long a stall can last before the viewer notices,
        // and roughly what LOW gives back in click-to-first-frame. They bracket PRD §9's
        // two-second first-frame bar by reasoning, not by measurement, and want profiling on
        // the low-end baseline before anyone calls them tuned.
        val cacheSecs: String
        val readaheadSecs: String
        when (profile) {
            BufferingProfile.LOW -> {
                cacheSecs = "2"
                readaheadSecs = "1"
            }
            BufferingProfile.BALANCED -> {
                cacheSecs = "10"
                readaheadSecs = "5"
            }
            BufferingProfile.GENEROUS -> {
                cacheSecs = "30"
                readaheadSecs = "20"
            }
        }
        mpv.setOption("cache", "yes")
        mpv.setOption("cache-secs", cacheSecs)
        mpv.setOption("demuxer-readahead-secs", readaheadSecs)
    }

    private fun applyHeaders(
        mpv: MpvClient,
        request: StreamRequest,
    ) {
        request.userAgent?.let { mpv.setOption("user-agent", it) }

        // -append, one header per call, rather than a comma-joined list.
        // `http-header-fields` is a comma-separated list option, so a value containing a
        // comma — as a Cookie routinely does — would silently split into two malformed
        // headers. Appending sidesteps the escaping question entirely.
        request.headers.forEach { header ->
            mpv.setOption("http-header-fields-append", "${header.name}: ${header.value}")
        }
    }

    private fun observeDefaults(mpv: MpvClient) {
        // STRING because mpv renders flags as "yes"/"no" in that format, which keeps the JNI
        // payload a plain string and the C side free of node-tree walking.
        listOf("pause", "paused-for-cache", "core-idle", "seekable").forEach {
            mpv.observeProperty(it, MpvClient.Format.STRING)
        }
        // NONE: notification with no payload. track-list is a node, and re-reading its flat
        // sub-properties on change is cheaper than marshalling the tree across JNI.
        mpv.observeProperty("track-list", MpvClient.Format.NONE)
    }

    private fun onEvent(event: MpvEvent) {
        if (released.get()) return
        val mpv = client ?: return

        when (event.eventId) {
            MpvClient.EventId.LOG_MESSAGE -> {
                val text = event.value ?: return
                MpvLog.mpvSaid(event.name, text)
                recentDiagnostics.add(MpvLogRedaction.redact(text))
            }

            MpvClient.EventId.START_FILE -> {
                flags = MpvStateDerivation.Flags()
                publish(MpvStateDerivation.stateFor(flags))
            }

            MpvClient.EventId.FILE_LOADED -> {
                flags = flags.copy(fileLoaded = true)
                refreshTracks(mpv)
                refreshFlags(mpv)
            }

            MpvClient.EventId.PLAYBACK_RESTART -> refreshFlags(mpv)

            MpvClient.EventId.PROPERTY_CHANGE -> onPropertyChange(mpv, event)

            MpvClient.EventId.END_FILE -> onEndFile(event)

            else -> Unit
        }
    }

    private fun onEndFile(event: MpvEvent) {
        val error =
            MpvErrorMapping.endFileError(
                reason = event.endFileReason,
                errorCode = event.endFileError,
                diagnostic = recentDiagnostics.snapshot(),
            )
        when {
            error != null -> publish(PlaybackState.Failed(error))
            event.endFileReason == MpvErrorMapping.EndFileReason.EOF -> publish(PlaybackState.Ended)
            // STOP/QUIT/REDIRECT: our own teardown, or mpv following a redirect. Neither is
            // something the viewer should see a state change for.
            else -> Unit
        }
    }

    private fun onPropertyChange(
        mpv: MpvClient,
        event: MpvEvent,
    ) {
        when (event.name) {
            "seekable" -> _isSeekable.value = MpvStateDerivation.flagOf(event.value)
            "track-list" -> refreshTracks(mpv)
            "pause" -> updateFlags(flags.copy(pause = MpvStateDerivation.flagOf(event.value)))
            "paused-for-cache" -> updateFlags(flags.copy(pausedForCache = MpvStateDerivation.flagOf(event.value)))
            "core-idle" -> updateFlags(flags.copy(coreIdle = MpvStateDerivation.flagOf(event.value)))
            else -> Unit
        }
    }

    private fun updateFlags(next: MpvStateDerivation.Flags) {
        flags = next
        publish(MpvStateDerivation.stateFor(next))
    }

    private fun refreshFlags(mpv: MpvClient) {
        updateFlags(
            flags.copy(
                pause = MpvStateDerivation.flagOf(mpv.getProperty("pause")),
                pausedForCache = MpvStateDerivation.flagOf(mpv.getProperty("paused-for-cache")),
                coreIdle = MpvStateDerivation.flagOf(mpv.getProperty("core-idle")),
            ),
        )
        _isSeekable.value = MpvStateDerivation.flagOf(mpv.getProperty("seekable"))
    }

    private fun refreshTracks(mpv: MpvClient) {
        val count = mpv.getProperty("track-list/count")?.toIntOrNull() ?: return
        val rows =
            (0 until count).mapNotNull { index ->
                val id = mpv.getProperty("track-list/$index/id")?.toIntOrNull() ?: return@mapNotNull null
                val type = mpv.getProperty("track-list/$index/type") ?: return@mapNotNull null
                MpvTrackMapping.RawTrack(
                    id = id,
                    type = type,
                    title = mpv.getProperty("track-list/$index/title"),
                    lang = mpv.getProperty("track-list/$index/lang"),
                    selected = MpvStateDerivation.flagOf(mpv.getProperty("track-list/$index/selected")),
                )
            }
        _tracks.value = MpvTrackMapping.toTrackSelection(rows)
    }

    private fun publish(next: PlaybackState) {
        val previous = _state.value
        if (previous == next) return
        // Failed latches. The contract says a failed engine is spent and the shell disposes
        // it, so a late property change must not resurrect it into Playing.
        //
        // Only Failed — not Ended. The contract's own `isTerminal` also counts Ended, which is
        // right for the shell (it disposes on both), but a live stream can emit END_FILE with
        // EOF and be reloaded by mpv, and latching Ended here would freeze an engine whose
        // source recovered by itself.
        if (previous is PlaybackState.Failed) return
        MpvLog.transition(previous, next)
        _state.value = next
    }
}

File: apps/tvos/Packages/PlayerMPV/Sources/PlayerMPV/MPVEngine.swift
// SPDX-FileCopyrightText: 2026 Spidola contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

import Foundation
import Libmpv
import OSLog
import PlayerContract
import SwiftUI

/// The MPVKit engine — the tvOS default (TECH_SPEC §8, §14: AVPlayer's format range is too narrow
/// to be the default for IPTV reality, so mpv's codec breadth takes the position).
///
/// Construction is deliberately free of I/O — no `mpv_create`, no session, no thread. The zap path
/// destroys and rebuilds engines on every channel change and its budget is an acceptance test
/// (PRD §9), so everything expensive waits for `load`.
@MainActor
public final class MPVEngine: PlaybackEngine {
  public nonisolated let id = EngineID.mpv
  public nonisolated let states: AsyncStream<PlaybackState>

  public private(set) var tracks = TrackSelection()
  public private(set) var isSeekable = false

  private let statesContinuation: AsyncStream<PlaybackState>.Continuation

  /// The render target. Built at init because it is cheap (no drawable is allocated until mpv
  /// renders) and because mpv needs its address before `mpv_initialize` — and held here, strongly,
  /// because MPVKit's MoltenVK context bridges `wid` back to a `CAMetalLayer` **unretained**. If
  /// this reference dies while the core lives, mpv renders into freed memory.
  ///
  /// Holding it here is necessary but not sufficient: the core outlives this engine by design, on
  /// `MPVCoreDisposal`'s thread, so the layer is handed to that thread as well and released only
  /// once `mpv_destroy` has returned.
  private let metalLayer = MPVMetalLayer()

  private let audioSession = MPVAudioSession()
  private let nowPlaying = MPVNowPlaying()

  /// The core handle. Owned exclusively by the main actor; `nil` once `stop()` has handed it off.
  private var core: MPVHandle?
  private var tasks: [Task<Void, Never>] = []

  private var facts = MPVPlaybackFacts()
  private var currentState: PlaybackState = .idle
  /// The last failure mpv's log implied. mpv reports *why* a load failed in its log stream and
  /// *that* it failed in a later `MPV_EVENT_END_FILE`, so the classification has to be carried
  /// across the gap; see `MPVErrorMapping.logHint`.
  private var lastLogHint: EngineError?
  private var isStopped = false
  /// Whether the viewer was playing when the system interrupted us, so `.ended(shouldResume:)`
  /// resumes only what it actually paused.
  private var wasPlayingBeforeInterruption = false
  private var mediaTitle: String?

  public init() {
    let (stream, continuation) = AsyncStream<PlaybackState>.makeStream(bufferingPolicy: .unbounded)
    self.states = stream
    self.statesContinuation = continuation
    continuation.yield(.idle)
  }

  public func makeSurface() -> AnyView {
    AnyView(MPVMetalSurface(metalLayer: metalLayer))
  }

  // MARK: - Loading

  public func load(_ request: StreamRequest) {
    guard !isStopped, core == nil else { return }

    Logger.mpv.info(
      """
      loading \(MPVRedaction.locatorSummary(request.locator), privacy: .public) \
      headers=[\(MPVRedaction.headerNames(request.headers), privacy: .public)] \
      user-agent=\(MPVRedaction.userAgentPresence(request.userAgent), privacy: .public) \
      buffering=\(request.buffering.rawValue, privacy: .public)
      """)

    emit(.loading)
    do {
      let core = try makeCore(for: request)
      self.core = core
      try startEventLoop(on: core)
      startSystemIntegration()
      try core.command(["loadfile", request.locator, "replace"])
    } catch let error as MPVCallError {
      Logger.mpv.error(
        "load failed during \(error.operation, privacy: .public): \(error.code, privacy: .public)")
      emit(.failed(MPVErrorMapping.engineError(mpvError: error.code, logHint: lastLogHint)))
    } catch {
      emit(.failed(.unknown(detail: "unexpected engine error")))
    }
  }

  /// Builds and initialises the core.
  ///
  /// Everything here happens **before** `mpv_initialize` because mpv only reads these at
  /// initialisation: `wid` and the video-output stack cannot be changed afterwards, and setting
  /// them late silently leaves mpv on its own default output with nowhere to draw.
  private func makeCore(for request: StreamRequest) throws -> MPVHandle {
    let core = try MPVHandle.create()
    do {
      try configure(core, for: request)
    } catch {
      // The handle exists but nothing owns it yet: `self.core` is only assigned once this returns,
      // so an early throw here would strand the core with no path to `stop()` and leak a decoder
      // on every failed load — which on a bad source is every zap.
      MPVCoreDisposal.dispose(core.raw, keepingAlive: metalLayer)
      throw error
    }
    return core
  }

  private func configure(_ core: MPVHandle, for request: StreamRequest) throws {
    // The render path. See `MPVMetalLayer` for why this is `wid` + MoltenVK rather than libmpv's
    // render API — in short, that API has no Metal type and tvOS has no future in OpenGL.
    try core.setWindowID(Unmanaged.passUnretained(metalLayer).toOpaque())
    try core.setOption("vo", "gpu-next")
    try core.setOption("gpu-api", "vulkan")
    try core.setOption("gpu-context", "moltenvk")

    // VideoToolbox is the whole reason mpv is viable on this hardware — an Apple TV cannot software
    // decode 4K HEVC. The simulator has no VideoToolbox backing, so it decodes in software there;
    // without this branch every simulator run fails at the decoder rather than showing a picture.
    #if targetEnvironment(simulator)
      try core.setOption("hwdec", "no")
    #else
      try core.setOption("hwdec", "videotoolbox")
    #endif

    // mpv would otherwise take the terminal and stdin, neither of which exists in an app, and its
    // config/OSD would fight the shell's own UI (PRD §6.3: the shell owns the overlays).
    try core.setOption("config", "no")
    try core.setOption("terminal", "no")
    try core.setOption("input-default-bindings", "no")
    try core.setOption("input-vo-keyboard", "no")
    try core.setOption("osd-level", "0")
    // Rotation metadata is the source's business, not ours to re-apply on a TV.
    try core.setOption("video-rotate", "no")

    // Deliberately absent: `osc`. mpv registers it only under `#if HAVE_LUA`, and MPVKit builds
    // Lua for macOS alone (`-Dlua=luajit`) while every other platform, tvOS included, gets
    // `-Dlua=disabled`. Setting it here would throw MPV_ERROR_OPTION_NOT_FOUND and fail every
    // load on device while passing on a Mac. There is no on-screen controller to disable without
    // Lua, so the option has nothing to do here anyway.

    try apply(request: request, to: core)
    for option in MPVOptions.aspectOptions(for: .fit) {
      try core.setOption(option)
    }

    try core.initialize()
  }

  private func apply(request: StreamRequest, to core: MPVHandle) throws {
    for option in MPVOptions.cacheOptions(for: request.buffering) {
      try core.setOption(option)
    }
    // Verbatim through a node array — mpv's option parser would split a header value on its commas.
    try core.setStringList("http-header-fields", MPVOptions.headerFields(request.headers))
    if let userAgent = request.userAgent {
      try core.setOption("user-agent", userAgent)
    }
  }

  /// The properties whose changes drive the state machine.
  ///
  /// `track-list` is observed with `MPV_FORMAT_NONE` — a bare "it changed" notification. The value
  /// is then fetched as JSON, because reading a node tree through the event payload would mean
  /// hand-rolled pointer traversal for data `MPVTrackList` can parse from a string.
  private static let observedProperties: [MPVObservedProperty] = [
    MPVObservedProperty(name: "core-idle", format: MPV_FORMAT_FLAG),
    MPVObservedProperty(name: "paused-for-cache", format: MPV_FORMAT_FLAG),
    MPVObservedProperty(name: "eof-reached", format: MPV_FORMAT_FLAG),
    MPVObservedProperty(name: "pause", format: MPV_FORMAT_FLAG),
    MPVObservedProperty(name: "idle-active", format: MPV_FORMAT_FLAG),
    MPVObservedProperty(name: "seekable", format: MPV_FORMAT_FLAG),
    MPVObservedProperty(name: "duration", format: MPV_FORMAT_DOUBLE),
    MPVObservedProperty(name: "time-pos", format: MPV_FORMAT_DOUBLE),
    MPVObservedProperty(name: "media-title", format: MPV_FORMAT_STRING),
    MPVObservedProperty(name: "track-list", format: MPV_FORMAT_NONE),
  ]

  private func startEventLoop(on core: MPVHandle) throws {
    let clientHandle = try core.createWeakClientHandle(name: "spidola_events")
    let events = MPVEventLoop.start(
      clientHandle: clientHandle, observedProperties: Self.observedProperties)

    // The stream is hoisted out of the closure deliberately: `AsyncStream` is `Sendable`, so the
    // task captures only it and a weak `self`, never the engine's non-`Sendable` internals.
    tasks.append(
      Task { [weak self] in
        for await event in events {
          guard let self else { return }
          self.handle(event)
        }
      })
  }

  private func startSystemIntegration() {
    audioSession.activate()
    nowPlaying.activate()

    let interruptions = audioSession.interruptions
    let commands = nowPlaying.commands

    tasks.append(
      Task { [weak self] in
        for await interruption in interruptions {
          guard let self else { return }
          self.handle(interruption)
        }
      })
    tasks.append(
      Task { [weak self] in
        for await command in commands {
          guard let self else { return }
          self.handle(command)
        }
      })
  }

  // MARK: - Transport

  public func play() {
    setPause(false)
  }

  public func pause() {
    setPause(true)
  }

  private func setPause(_ paused: Bool) {
    guard let core else { return }
    do {
      try core.setFlag("pause", paused)
    } catch {
      Logger.mpv.error("failed to set pause=\(paused, privacy: .public)")
    }
  }

  public func seek(toSeconds seconds: Double) {
    // The contract makes this a no-op rather than an error when unseekable, so the caller need not
    // guard — live streams are the common case and every call site would otherwise repeat this.
    guard isSeekable, let core else { return }
    do {
      try core.command(["seek", String(seconds), "absolute"])
    } catch {
      Logger.mpv.error("seek failed")
    }
  }

  public func select(track: TrackID) {
    guard let core, let match = tracks.available.first(where: { $0.id == track }) else { return }
    do {
      // mpv splits track selection across two properties by kind; the contract has one call, so the
      // kind we already recorded picks the property.
      switch match.kind {
      case .audio: try core.setProperty("aid", track.rawValue)
      case .subtitle: try core.setProperty("sid", track.rawValue)
      }
    } catch {
      Logger.mpv.error("track selection failed")
    }
  }

  public func clearSubtitle() {
    guard let core else { return }
    do {
      // "no" is mpv's off value for a track property — distinct from an id, which is why the
      // contract keeps this separate from `select`.
      try core.setProperty("sid", "no")
    } catch {
      Logger.mpv.error("clearing subtitle failed")
    }
  }

  public func setAspect(_ mode: AspectMode) {
    guard let core else { return }
    do {
      for option in MPVOptions.aspectOptions(for: mode) {
        try core.setProperty(option.name, option.value)
      }
    } catch {
      Logger.mpv.error("aspect change failed")
    }
  }

  // MARK: - Teardown

  /// Tears the engine down. Idempotent, per the contract: the shell calls this on dispose and on
  /// the terminal-state path, and neither knows whether the other ran.
  ///
  /// **The ordering, which is the delicate part.** A use-after-free here is a crash on every zap, so
  /// the design removes the race rather than narrowing it. Two handles exist, and each is touched by
  /// exactly one thread for its whole life:
  ///
  /// - `core` — strong, main actor only.
  /// - the event loop's weak client — its own thread only, created and destroyed there.
  ///
  /// Because no handle is shared, no thread can free one another is inside. What remains is waking
  /// the event thread, which sits blocked in `mpv_wait_event`. We never poke it: destroying the last
  /// **strong** handle makes libmpv send `MPV_EVENT_SHUTDOWN` to every **weak** client, which is the
  /// wake-up, delivered by the core itself. The event thread then destroys its own handle and exits.
  ///
  /// `mpv_destroy` on the last strong handle blocks until the weak clients respond and the core
  /// finishes uninitialising, which is why it runs on a disposal thread and not here — blocking the
  /// main actor is what the zap budget cannot afford. This engine sets `core = nil` before handing
  /// the pointer over, so the main actor's last use of it precedes the hand-off by construction.
  ///
  /// The render target crosses with it, retained past `mpv_destroy`. mpv still holds it —
  /// unretained, as a bare `wid` address — for as long as it is uninitialising, while this engine's
  /// own reference to it can die the instant this method returns: on the `deinit` path below, that
  /// is the next thing that happens.
  public func stop() {
    guard !isStopped else { return }
    isStopped = true

    guard let core else {
      statesContinuation.finish()
      return
    }
    self.core = nil

    for task in tasks { task.cancel() }
    tasks.removeAll()
    nowPlaying.deactivate()
    audioSession.deactivate()

    Logger.mpv.info("engine stopping")
    MPVCoreDisposal.dispose(core.raw, keepingAlive: metalLayer)
    statesContinuation.finish()
  }

  /// The contract's backstop for a dropped reference. Without it a live core keeps decoding with
  /// no owner left to stop it — the event task holds its stream, mpv holds the decoder, and the
  /// process-wide audio session and now-playing info stay claimed. `isolated deinit` runs on the
  /// main actor, so this is the same idempotent `stop()` the shell calls, not a second teardown
  /// path.
  isolated deinit {
    stop()
  }

  // MARK: - Events

  private func handle(_ event: MPVEvent) {
    switch event {
    case .shutdown:
      // Expected during our own teardown, and already handled there. Unprompted, it means the core
      // died under us — terminal, and the shell must be told rather than left on a frozen frame.
      guard !isStopped else { return }
      emit(.failed(.unknown(detail: "mpv core shut down unexpectedly")))

    case .fileLoaded:
      facts.fileLoaded = true
      refreshTracks()
      reduce()

    case .endFile(let reason, let mpvError):
      // `nil` means an end-of-file we caused (our own stop, a redirect); saying nothing is correct.
      guard
        let outcome = MPVErrorMapping.endFileOutcome(
          reason: reason, mpvError: mpvError, logHint: lastLogHint)
      else { return }
      emit(outcome)

    case .propertyChanged(let name, let value):
      handleProperty(name: name, value: value)

    case .logMessage(_, let text):
      // Classified and dropped. The text never reaches OSLog — mpv logs full stream URLs, and an
      // Xtream URL carries the account in its path (TECH_SPEC §12). Only the derived class, which
      // is one of six fixed cases, survives this line.
      if let hint = MPVErrorMapping.logHint(from: text) {
        lastLogHint = hint
      }
    }
  }

  private func handleProperty(name: String, value: MPVPropertyValue) {
    switch (name, value) {
    case ("core-idle", .flag(let flag)):
      facts.coreIdle = flag
    case ("paused-for-cache", .flag(let flag)):
      facts.pausedForCache = flag
    case ("eof-reached", .flag(let flag)):
      facts.eofReached = flag
    case ("pause", .flag(let flag)):
      facts.paused = flag
    case ("idle-active", .flag(let flag)):
      // mpv going idle after a file means the playlist ran out. `eof-reached` covers the normal
      // run-out; this catches the case where mpv drops to idle without setting it.
      if flag && facts.fileLoaded { facts.eofReached = true }
    case ("seekable", .flag(let flag)):
      isSeekable = flag
    case ("duration", .double), ("time-pos", .double):
      updateNowPlaying()
      return
    case ("media-title", .string(let title)):
      mediaTitle = title
      updateNowPlaying()
      return
    case ("track-list", _):
      refreshTracks()
      return
    default:
      return
    }
    reduce()
  }

  private func refreshTracks() {
    guard let core, let json = core.string("track-list") else { return }
    tracks = MPVTrackList.parse(json: json)
  }

  /// Re-derives the contract state from the facts and publishes it if it moved.
  ///
  /// De-duplicating matters: mpv fires these properties independently and several arrive per
  /// second, most of which do not change the derived state. Yielding each one would turn the state
  /// stream into a firehose the shell would have to de-duplicate itself.
  private func reduce() {
    emit(MPVStateReducer.state(from: facts))
    updateNowPlaying()
  }

  private func updateNowPlaying() {
    guard let core else { return }
    nowPlaying.update(
      MPVNowPlayingState(
        title: mediaTitle,
        duration: core.double("duration"),
        position: core.double("time-pos"),
        rate: facts.paused ? 0 : 1,
        isLive: !isSeekable))
  }

  private func emit(_ state: PlaybackState) {
    guard !isStopped, state != currentState else { return }
    currentState = state
    Logger.mpv.info("state -> \(String(describing: state), privacy: .public)")
    statesContinuation.yield(state)
    if state.isTerminal { statesContinuation.finish() }
  }

  // MARK: - System integration

  private func handle(_ interruption: MPVAudioInterruption) {
    switch interruption {
    case .began:
      wasPlayingBeforeInterruption = currentState == .playing
      pause()
    case .ended(let shouldResume):
      // Both conditions are needed. The system's opinion alone would resume a channel the viewer
      // had paused before Siri ever arrived; ours alone would talk over whatever the interruption
      // left playing.
      guard shouldResume, wasPlayingBeforeInterruption else { return }
      play()
    case .routeChanged:
      updateNowPlaying()
    }
  }

  private func handle(_ command: MPVTransportCommand) {
    switch command {
    case .play: play()
    case .pause: pause()
    case .togglePlayPause:
      if facts.paused { play() } else { pause() }
    case .seek(let seconds): seek(toSeconds: seconds)
    }
  }
}
